### PR TITLE
Thing "actions" config param handling made generic

### LINKED ISSUE #1491: [thing-details] Extend special handling of "actions" parameter group to all bindings
## The problem
Currently, the Thing Details view handles Z-Wave binding special, and presents a list of <b>actions</b> (`<parameter-group name="actions">`) as clickable buttons, with confirmation prompt. 

This behavior is nice and desirable, but **specific to Z-Wave binding** only:

- https://github.com/openhab/openhab-webui/blob/60fd5f3c4636764c1b5908b542f601a95fc6af48/bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue#L374-L378

- https://github.com/openhab/openhab-webui/blob/60fd5f3c4636764c1b5908b542f601a95fc6af48/bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue#L110-L119

## Your suggestion
Make the abovementioned behavior more **generic** (remove hard-code for Z-Wave, allow **other bindings** to use it).

### Exemplary proposal
Use the "action" behavior IFF the following heuristic match is true:
`groupName == 'actions' && groupLabel === 'Actions' && groupDescription === 'Actions' && action.type === 'BOOLEAN'`
^--- The values have been picked to be backwards-compatible with [existing Zwave binding's implementation](https://github.com/openhab/org.openhab.binding.zwave/blob/6fe69d4b2267cf8657e4c437e505e2eb4151a18b/src/main/java/org/openhab/binding/zwave/internal/ZWaveConfigProvider.java#L195). Assumption is these values would never be i18n'ed and we want to avoid extra conditions. 

_Note:_ A bigger (more semantically correct) change may include changing the Configuration Description DSL to include a new parameter group and/or type (say: `action`) and remove the heuristic matching by group/param name. This would be a OH-core change, though and given this Z-Wave hardcode has been around for ~3y now, seems to be an overkill to introduce now IMHO.

### Rationale
It is customary for some devices to include a "_reset to factory default_" functionality. Z-Wave binding's paradigm of showing such one-off "commands" as actions seems appropriate and would create consistent UX.

## Your environment

```yaml
runtimeInfo:
  version: 3.4.0
```

## Additional information
_Disclaimer:_ I'm currently developing a binding which would benefit from this functionality.

### PR DESCRIPTION
**Pull request by mbronk**
For Main UI's Thing-Details view.
Fixes #1491

Signed-off-by: Mateusz Bronk <bronk.m+gh@gmail.com>

### CODE REVIEW

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (ghys) [inline bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue:115]:
```suggestion
              <f7-list-button :color="action.verify ? 'yellow' : 'blue'" v-for="action in configActions" :key="action.name" :title="action.label" @click="doConfigAction(action)" :tooltip="action.description" />
```

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (ghys) [inline bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue:375]: This heuristic is good enough but since parameter groups have a "context" attribute:

https://next.openhab.org/javadoc/latest/org/openhab/core/config/core/dto/configdescriptionparametergroupdto

I'm wondering whether this could perhaps be the way to go - look for e.g. an "actions" context in the parameter group definition instead of checking for labels & descriptions - it certainly would be more elegant IMO.

We could still make an exception for the Z-Wave things unless @cdjackson agrees to add the "actions" context for the actions parameter group for them.

I'm not sure how easy that is to add such contexts to parameter groups, whether in declarative or imperative form.

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (ghys) [inline bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue:453]:
```suggestion
        `${action.description ?? action.label} ? ${action.verify ? '<p><small><strong class="text-color-yellow">WARNING:</strong>&nbsp;This action may be dangerous!</small></p>' : ''}`,
```
I'm not sure I understand this...

**Review by ghys — CHANGES_REQUESTED** [review]
Thanks for the implementation, some remarks (and further discussion) below!

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (mbronk) [reply bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue:115]: commited as-is (will squash later) - or feel free to squash if/when merging

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (mbronk) [reply bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue:375]: Indeed, there seems to be a "param group context" defined, though not documented too widely. Per [this](context) it would seem it may be a freeform string (the per-param context seems to be a pre-defined list).
Let me try to see how easy it is to set/use and report back (may take me about a week to do so, though)...

I 100% agree it would be more elegant than this heuristic.
That said, I think the ROI is there when ZWave binding also adopts it, otherwise we're back into heuristic-land :). Not sure what'd be the timeline, but how about a 2-step approach? I.e. this heuristic *for now* (possibly with an "or" for a by-context match) and follow through with a proper update if/once ZWave binding changes?

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (mbronk) [reply bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue:453]: Oh. and the abovementioned code was not a ternary IF statement! It is just a good old questionmark (to be rendered as-is in the output :) ).

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (ghys) [reply bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue:375]: > I.e. this heuristic for now (possibly with an "or" for a by-context match) and follow through with a proper update if/once ZWave binding changes?

Works for me!

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (ghys) [reply bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue:453]: > The "actions" behaviour of Main UI does ask for a confirmation for all the actions, of course... so I figured to bring back at least some of the "verify" semantics, and add a notion of "hey, the binding developer considered this dangerous".

Wow I completely forgot this "verify" attribute existed... 🤯

Fully agree with your assessment - this particular line of code I considered quite "convoluted" though with the muliple ternary expressions. If I understand it right it does display the description or label, or when there's none, displays a warning message *if* the verify flag is true? Is that your intent?

Comment (mbronk) [reply bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue:453]: > or when there's none....

As per my 2nd comment... the `?` here is not a ternary expresssion, actually, rather the same questionmark that existed in the original. The intent is to:
  1.  Display description (and fallback to label if none)
  1. (orthogonal to the above): If `verify==true` --> add an extra warning.

I now see how poorly it reads code-wise (it was much better in IDE, with coloring, etc.). Unfortunately, I have literally never written anything in vue, so while I assume using some templated text/conditional controls would make it more readable, it would take me some extra learning to find the syntax out. Do you happen to have any tips in particular?

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (mbronk) [reply bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue]: The "by-context" method does work indeed, so I added it as primary.
Any param group constructed with `.withContext("actions")` will now be picked as action container.

As per our discussion, I've left the heuristic match by `name` and `label` for compat with today's Z-wave binding. 
Note I had to pull the "by description" criterion, since Z-Wave doesn't use it uniformly and only sets it on non-controller devices...

--
While at it, I attempted to make it a bit more generic (hopefully the code explains the intent well enough). Do ask if it raises any concerns.

File: bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue
Comment (mbronk) [reply bundles/org.openhab.ui/web/src/pages/settings/things/thing-details.vue]: Extracted to separate variables - hope it'll be more readable now.

**Comment by ghys** [comment]
@mbronk Please review the ESLint errors: https://github.com/openhab/openhab-webui/actions/runs/3165087894/jobs/5209020387
(you can also see _some_ in the Files tab but there's a limit on how many can be displayed per file, so some are missing towards the end of the file in that view).

If you can, use an editor that can highlight ESLint errors, for instance VS Code with the ESLint plugin - it's easier!

**Comment by mbronk** [comment]
@ghys Yeah, sorry about that! 
I don't have any git hooks set up for this repo, and completely forgot to run `npm run lint` manually before pushing... Since the CI only runs after manually approved for 1st time committers... I then have forgotten about it :/ 

Anyways, eslint should be happy now. Could you please approve another CI build?

**Review by ghys — APPROVED** [review]
Thanks - LGTM!
I just removed the tooltip with the description on the action buttons because it didn't look too good on mobile IMO, especially with potentially long descriptions. You have that description on the confirmation dialog anyway so it's probably enough...?
Nice new feature!

(build failures were temporary, the build actually succeeded: checks/webpack-stats are green)

**Pull request merged**
